Fix up symbol naming checks so --validate-goto-model can be enabled in regression testing.

chrisr-diffblue opened the pull request:
This PR supersedes #3661 by taking the first commit in that PR (which makes regression tests run with --validate-goto-program) and then the subsequent commits in this PR fix up a few places that causes regression tests to fail. The final commit in this PR adds a few special cases to the `symbol.is_well_formed()` check that was previously added in #3193.

I have a few concerns about these additional 'special cases' - so I'd welcome feedback on those - they feel a bit like a breaking of abstraction.

@kroening Feel free to either cherry-pick commits from this PR into #3661, or close that and focus on this PR instead. As mentioned by @tautschnig - this PR also supersedes/compliments #2734 as well.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
enable --validate-goto-model for all cbmc regressions

This will prevent the introduction of changes that violate the checks
in the goto-validation procedure.

chrisr-diffblue committed:
Fix base_name of return' symbol to match its name

chrisr-diffblue committed:
Fix base_name of argc' symbol to match its name

chrisr-diffblue committed:
Ensure the gcc_builtin_va_arg symbol has a mode

chrisr-diffblue committed:
Make #anon symbols have consistent naming between base_name and name

chrisr-diffblue committed:
Ensure the '$exception_flag' symbol has a valid mode

chrisr-diffblue committed:
Refine symbol well-formedness checks to handle special cases

The symbol naming check added in df616e4 attempts to assert that the 'base_name'
of a symbol is a suffix of that symbol's 'name'. In general, this should be the
case, but there are certain classes of symbols where this does not currently
hold. This commit adds special cases to handle:

* Type symbols that are typedefs or tags
* Symbols that have been renamed by the linker

With these special cases handled, it is possible to run the CORE regression
tests with --validate-goto-model enabled.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 25b53393b71246d077df89c33cb11f1fa7329ecf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97112409

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 94da2fcc42b4eb09a5012c1674109e5a3a24913c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97410307

tautschnig commented:
Just noting that this will help us close out #2734, where I complained about the missing test.

chrisr-diffblue commented:
Good spot!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 95bfb7d4ce92ffbc3348595f6ca3ece98240edb7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97418621

tautschnig commented on regression/cbmc/CMakeLists.txt:
```diff
@@ -1,3 +1,3 @@
 add_test_pl_tests(
-    "$<TARGET_FILE:cbmc>" -X smt-backend
+    "$<TARGET_FILE:cbmc> --validate-goto-model" -X smt-backend
```
What about `--validate-ssa`?

chrisr-diffblue replied:
I'm just extending #3661 to fix the stuff that was failing due to df616e4 in the interest of getting PRs merged, I'd suggest it might be better to enable `--validate-ssa` in a separate PR.

tautschnig commented on regression/cbmc/Float-zero-sum1/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
This test (and several others) is marked `KNOWNBUG`, then a number of apparent bug fixes are made, but no test is re-enabled? Is that intentional?

chrisr-diffblue replied:
Thats a verbatim cherry-pick from @kroening in #3661 - I did try re-enabling these tests but the failed validation checks for reasons other than this particular symbol naming check, so I felt it was better to focus this PR on just the symbol naming check - mostly to work towards progressing #3661

tautschnig replied:
Ok, we just need to make sure we get those tests back into the set as soon as possible.

smowton replied:
Would it be better to remove the checks that don't currently hold across the board? Those checks seem less likely to be missed in the future than a disable test?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4c699cf014f892f5a25c6c151332b02e6a01d611).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97425148

chrisr-diffblue committed:
Correct validation check for return statements from void functions

chrisr-diffblue commented:
I'm going to mark this back as "in progress" "do not merge" and I'll work on getting the rest of the checks fixed up - I've made some progress already, getting the list of disabled tests down to about 4 or 5, just need to tackle the remaining ones.

chrisr-diffblue committed:
Correct symbol table type check in cases where return removal has modified types

Return removal sets the return type of a function symbol table entry to 'void',
but some callsites still expect the original type (e.g. if a function is passed
as a parameter to a different function)

chrisr-diffblue committed:
Make goto_programt::instructiont::validate() accept incomplete array types

chrisr-diffblue committed:
Make goto_programt::instructiont::validate() accept differing return types

When a function declaration and its definition appear in different translation
units they may have different return types. This commit relaxes the checks
in goto_programt::instructiont::validate() in these circumstances.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6ce0915f38fb89e86be056db5868583d5a4dbc85).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98645639
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a6b31234e6d5623029b2c231320d0f92c9380700).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98657951
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

chrisr-diffblue commented:
@tautschnig @smowton @kroening This is now ready for re-review - in particular, there are now no regression tests being disabled by this PR. Apologies for the delay getting this PR ready for re-review...

tautschnig commented on src/goto-programs/goto_program.cpp:
```diff
@@ -14,6 +14,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <ostream>
 #include <iomanip>
 
+#include <goto-programs/remove_returns.h>
```
I'd prefer `#include "remove_returns.h"` and placing that at the bottom of the list of includes, for consistency with how that's done in other files.

And a nit-pick: s/differnet/different/ in the commit message.

chrisr-diffblue replied:
Fixed.

tautschnig commented on src/goto-programs/goto_program.cpp:
```diff
@@ -709,13 +710,44 @@ void goto_programt::instructiont::validate(
+              // type of the function and use that full type to check
+              // consistency with the goto program
+              auto full_table_symbol_type = to_code_type(table_symbol->type);
+              full_table_symbol_type.return_type() = return_value_symbol->type;
```
This can be simplified by using `original_return_type` (from `remove_returns.{h,cpp}`).

chrisr-diffblue replied:
Thanks - I did actually look for a utility function to use here, but somehow managed to miss that. I much appreciate the pointer.

tautschnig commented:
Some smaller suggestions for improvement above, but otherwise I'd suggest to merge this as soon as possible.

chrisr-diffblue commented:
I'll wait for CI to pass, then merge if nobody else wants to block it.

smowton commented:
The function-type mismatch due to declaration vs. definition inconsistency smells a bit off. Is there future work to figure out why that results in a symbol-table vs. goto-program type clash?

chrisr-diffblue commented:
@smowton I don't have a fixed plan to dig into that deeper, but yes, I'd like to at some point - though its the kind of code that, say, GCC will compile relatively happily... `regression/cbmc/Linking7/return_type*.c` is an example of such a program.

tautschnig commented:
@smowton @chrisr-diffblue The goto-program linker permits various type inconsistencies to compile (poor) real-world C code. Some of it is fixed up by casts done in symbolic execution, other bits are fixed via `object_type_updates` in `link_goto_model.cpp`. There is a risk that some inconsistencies prevail, and we might have to become more stringent about the check that is being relaxed here. But as long as a type cast can fix it we're probably ok...

chrisr-diffblue commented:
OK, CI has passed, so I'm going to hit merge. If anyone feels strongly that I shouldn't have relaxed the check declaration vs. definition check, I can sort that out in a different PR.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 933c2ef117821f05879046b6fea74ada8c9b9333).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98841367
